main.go

```diff
@@ -0,0 +1,266 @@
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/aws/aws-lambda-go/events"
+	"github.com/aws/aws-lambda-go/lambda"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	log "github.com/sirupsen/logrus"
+	"go.mozilla.org/mozlogrus"
+)
+
+func init() {
+	if os.Getenv("CT_DEBUG_LOGGING") == "1" {
+		log.SetLevel(log.DebugLevel)
+	} else {
+		log.SetLevel(log.InfoLevel)
+	}
+
+	mozlogrus.Enable("cloudtrail-streamer")
+}
+
+var (
+	globalConfig Config
+)
+
+type Config struct {
+	awsKinesisStream string // Kinesis stream for event submission
+	awsKinesisRegion string // AWS region the kinesis stream exists in
+
+	awsKinesisBatchSize int // Number of records in a batched put to the kinesis stream
+
+	awsKinesisClient *kinesis.Kinesis
+
+	awsSession *session.Session
+
+	eventType string // Whether to use the S3 or SNS event handler. Default is S3.
+}
+
+func (c *Config) init() error {
+	c.awsKinesisStream = os.Getenv("CT_KINESIS_STREAM")
+	if c.awsKinesisStream == "" {
+		return fmt.Errorf("CT_KINESIS_STREAM must be set")
+	}
+
+	c.awsKinesisRegion = os.Getenv("CT_KINESIS_REGION")
+	if c.awsKinesisRegion == "" {
+		return fmt.Errorf("CT_KINESIS_REGION must be set")
+	}
+
+	c.awsKinesisBatchSize = 500
+	batchSize := os.Getenv("CT_KINESIS_BATCH_SIZE")
+	if batchSize != "" {
+		var err error
+		c.awsKinesisBatchSize, err = strconv.Atoi(batchSize)
+		if err != nil {
+			log.Fatalf("Error converting CT_KINESIS_BATCH_SIZE (%v) to int: %s", batchSize, err)
+		}
+
+		if c.awsKinesisBatchSize > 500 {
+			return fmt.Errorf("CT_KINESIS_BATCH_SIZE is set to a value greater than 500")
+		}
+	}
+
+	c.awsSession = session.Must(session.NewSession())
+
+	c.awsKinesisClient = kinesis.New(
+		c.awsSession,
+		aws.NewConfig().WithRegion(c.awsKinesisRegion),
+	)
+
+	c.eventType = "S3"
+	eventType := os.Getenv("CT_EVENT_TYPE")
+	if eventType != "" {
+		c.eventType = eventType
+	}
+	if c.eventType != "S3" && c.eventType != "SNS" {
+		return fmt.Errorf("CT_EVENT_TYPE is set to an invalid value, %s, must be either 'S3' or 'SNS'", eventType)
+	}
+
+	return nil
+}
+
+type CloudTrailFile struct {
+	Records []map[string]interface{} `json:"Records"`
+}
+
+func fetchLogFromS3(s3Client *s3.S3, bucket string, objectKey string) (*s3.GetObjectOutput, error) {
+	logInput := &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(objectKey),
+	}
+
+	object, err := s3Client.GetObject(logInput)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchKey:
+				log.Errorf("%s: %s", s3.ErrCodeNoSuchKey, aerr)
```

Comment:
is there a need for the switch here? would aerr in both cases contain enough info?